presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java

```diff
@@ -149,33 +149,34 @@ public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
         {
             PlanNode source = context.rewrite(node.getSource());
 
-            TupleDomain<Symbol> tupleDomain = fromPredicate(metadata, session, node.getPredicate(), types).getTupleDomain();
-
+            // fromPredicate can be a very expensive operation. We should only call it if it's actually necessary.
```

Comment:
In extreme case, most operations can become expensive, so I doubt this comment is that useful. I appreciate thorough explanation you did in the commit message though.